### PR TITLE
Incremental Mean Volume Checkpoint

### LINKED ISSUE #1070: Add Mean Estimator checkpointing
It would be useful if our mean estimator periodically saved off the last mean volume to disk.  If the estimate approaches our threshold but fails to converge to the target, the current code will just continue to iterate while killing process will lose the volume.  

For larger volume sizes this can often be after a long wait.  It is possible the volume is still useful at this point.  I'm also not see at this time how the expected target is derived...

While I have not tested, I believe we _should_ be able to use such a checkpoint as a starting point by converting into basis via the adjoint and using as `b_coef`.  As part of this issue we can check this feature.

### PR DESCRIPTION
**Pull request by garrettwrong**
Close #1070 by adding incremental checkpoint and max iteration limits.

Altered proposed solution in issue slightly.  Once I got in the code I noticed it is more effective to return and restore coef directly than convert to/from Volume.

Still considering options for testing this.

### CODE REVIEW

**Comment by garrettwrong** [comment]
@j-c-c , since I added a bunch of doc string stuff probably best to start working through the review.  Thanks

**Comment by j-c-c** [comment]
> @j-c-c , since I added a bunch of doc string stuff probably best to start working through the review. Thanks

Sounds good. I'll start taking a look today.

File: src/aspire/reconstruction/mean.py [hunk 1]
```diff
@@ -200,12 +200,36 @@ def conj_grad(self, b_coef, tol=1e-5, regularizer=0):
...
-                f"Delta {norm(b_coef - self.apply_kernel(xk))} (target {target_residual})"
+                f"[Iter {self.i}]: Delta {norm(b_coef - self.apply_kernel(xk))} (target {target_residual})"
             )
 
-        x, info = cg(operator, b_coef.flatten(), M=M, callback=cb, tol=tol, atol=0)
+            # Optional checkpoint
+            if (
+                self.checkpoint_iterations
+                and (self.i % self.checkpoint_iterations) == 0
+            ):
```
Comment (j-c-c) [inline src/aspire/reconstruction/mean.py]: Might be overkill for grabbing checkpoints, but you could also save off if `self.i  == self.maxiter` or `self.maxiter - 1` (whichever is the last before `cg` raises).

File: src/aspire/reconstruction/estimator.py [hunk 2]
```diff
@@ -35,6 +64,37 @@ def __init__(self, src, basis, batch_size=512, preconditioner="circulant"):
...
+                checkpoint_iterations = -1
+
+            if not checkpoint_iterations > 0:
+                raise ValueError(
+                    "`checkpoint_iterations` should be a positive integer or `None`."
+                )
+        self.checkpoint_iterations = checkpoint_iterations
+
+        # Create checkpointing dirs as needed
+        parent = Path(checkpoint_prefix).parent
```
Comment (j-c-c) [inline src/aspire/reconstruction/estimator.py]: Would probably be a rare case, but if `checkpoint_prefix` is `None` I think this breaks.

**Review by j-c-c — CHANGES_REQUESTED** [review]
Looks good. Just a couple things.

File: src/aspire/reconstruction/estimator.py (hunk 2 above)
Comment (garrettwrong) [reply src/aspire/reconstruction/estimator.py]: Yes, that's a bug, pushing a small fix, thanks.

File: src/aspire/reconstruction/mean.py (hunk 1 above)
Comment (garrettwrong) [reply src/aspire/reconstruction/mean.py]: Might as well. Added.

**Review by j-c-c — DISMISSED** [review]
Looks good!

File: src/aspire/reconstruction/estimator.py [hunk 3]
```diff
@@ -7,14 +9,40 @@
...
         Convolving a volume with this kernel is equal to projecting and backproject-ing that volume in each of the
         projection directions (with the appropriate amplitude multipliers and CTFs) and averaging over the whole
         dataset.
         Note that this is a non-centered Fourier transform, so the zero frequency is found at index 1.
+
+        :param src: `ImageSource` to be used for estimation.
+        :param basis: 3D Basis to be used during estimation.
+        :param batch_size: Optional batch size of images drawn from
+            `src` during back projection and kernel estimation steps.
+        :param preconditioner: Optional kernel preconditioner (`string`).
```
Comment (janden) [inline src/aspire/reconstruction/estimator.py:34]: Should mention that the two options here are `"circulant"` or `None`.

File: src/aspire/reconstruction/mean.py [hunk 4]
```diff
@@ -200,12 +200,42 @@ def conj_grad(self, b_coef, tol=1e-5, regularizer=0):
...
+                and (self.i % self.checkpoint_iterations) == 0
+            )
+            # or the last iteration when `maxiter` provided.
+            if self.maxiter:
+                _do_checkpoint |= self.i == (self.maxiter - 1)
+
+            # Optional checkpoint
+            if _do_checkpoint:
+                # Construct checkpoint path
+                path = f"{self.checkpoint_prefix}_iter{self.i}.npy"
```
Comment (janden) [inline src/aspire/reconstruction/mean.py]: Can we zero-prefix this (say `{self.i:04d}`) to make sorting easier?

**Review by janden — CHANGES_REQUESTED** [review]
Looks great! Just a few things.

File: src/aspire/reconstruction/estimator.py (hunk 3 above)
Comment (garrettwrong) [reply src/aspire/reconstruction/estimator.py:34]: Sure, added.

File: src/aspire/reconstruction/mean.py (hunk 4 above)
Comment (garrettwrong) [reply src/aspire/reconstruction/mean.py]: sure, added.

File: src/aspire/reconstruction/estimator.py (hunk 3 above)
Comment (janden) [reply src/aspire/reconstruction/estimator.py:34]: Thanks!

**Review by janden — APPROVED** [review]

**Pull request merged**